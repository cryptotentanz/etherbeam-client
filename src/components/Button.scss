@use '../styles/resources/sizes';
@use '../styles/resources/colors';
@use '../styles/resources/fonts';

.my-button,
.my-button-link {
  display: inline-flex;
  align-items: center;
  font-family: fonts.$primary;
  font-size: sizes.$size;
  padding: sizes.$spacer-xs;
  min-height: sizes.$size-xl;
}

.my-button,
.my-link-button {
  border-style: solid;
  border-width: sizes.$border;
  border-radius: sizes.$border-radius-s;
  font-weight: bold;

  .my-icon {
    &.my-positive {
      color: colors.$positive-action;
    }
  }
}
