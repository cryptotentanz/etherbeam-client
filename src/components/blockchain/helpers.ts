import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13
const UNISWAP_SWAP_URL = 'https://app.uniswap.org/#/swap'

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string | null, decimals: number, options?: AmountPrintOptions | null): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber | null, options?: AmountPrintOptions | null, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (
  amount: BigNumber | null,
  token: Token,
  options?: AmountPrintOptions | null
): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition | null => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}

export const getUniswapTradeUrl = (inputHash: string | null, outputHash: string | null): string => {
  return `${UNISWAP_SWAP_URL}?${inputHash ? `inputCurrency=${inputHash}&` : ''}${
    outputHash ? `outputCurrency=${outputHash}` : ''
  }`
}
