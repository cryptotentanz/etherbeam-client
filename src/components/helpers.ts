import { UserType } from 'libraries/types'
import { faUser, faUserShield, IconDefinition } from '@fortawesome/free-solid-svg-icons'

export const getUserTypeIcon = (type: UserType): IconDefinition => {
  switch (type) {
    case UserType.Administrator:
      return faUserShield
    default:
      return faUser
  }
}
