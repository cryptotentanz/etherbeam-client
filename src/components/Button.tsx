import React, { FC } from 'react'
import { IconProp } from '@fortawesome/fontawesome-svg-core'
import Icon from './Icon'
import './Button.scss'

type ColorProp = 'dark' | 'light' | 'primary' | 'secondary' | 'positive' | 'negative'
type SizeProp = 'm' | 'l' | 'xl'
type LabelBreakpointProp = 'xs' | 's' | 'm' | 'l'

interface ButtonProps {
  action?: () => void
  label?: string | null
  icon?: IconProp | null
  description?: string | null
  color?: ColorProp
  size?: SizeProp
  link?: boolean
  labelBreakpoint?: LabelBreakpointProp
  active?: boolean
  disabled?: boolean
  className?: string
}

const Button: FC<ButtonProps> = ({
  action,
  label,
  icon,
  description,
  color = 'light',
  size,
  link,
  labelBreakpoint,
  active,
  disabled,
  className,
}: ButtonProps) => {
  const getClassName = (): string => {
    const classNames = [link ? `my-button-link my-${color}-fg` : `my-button my-${color}-bg`, 'my-clickable']
    if (disabled) {
      if (!link) classNames.push('my-disabled-bg')
      classNames.push('my-disabled-fg')
    }
    if (active) classNames.push('my-active')
    if (className) classNames.push(className)

    return classNames.join(' ')
  }

  return (
    <button
      className={getClassName()}
      title={description ? description : undefined}
      onClick={action}
      disabled={disabled}
    >
      {icon ? (
        <Icon
          icon={icon}
          iconAlt={!link}
          labelBreakpoint={labelBreakpoint}
          label={label ? label : undefined}
          size={size}
        />
      ) : (
        <span className={size ? `my-size-${size}` : ''}>{label}</span>
      )}
    </button>
  )
}

export default Button
