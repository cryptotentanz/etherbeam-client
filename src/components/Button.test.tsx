import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { clearMocks, mockFontAwesome } from 'tests/helpers'
import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons'
import Button from './Button'

mockFontAwesome()

describe('<Button />', () => {
  const description = 'Description'
  const label = 'Label'
  const icon = faExclamationCircle
  const className = 'my-class'
  const action = jest.fn()

  let component: ShallowWrapper

  describe('Default', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} />)
    })

    it('should render button', () => {
      const button = component.find('button')
      expect(button.hasClass('my-button')).toBeTruthy()
      expect(button.hasClass('my-clickable')).toBeTruthy()
      expect(button.hasClass('my-light-bg')).toBeTruthy()
      expect(button.prop('onClick')).toBe(action)
    })

    it('should render icon', () => expect(component.find('span').at(0).prop('children')).toBe(label))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Description', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} description={description} />)
    })

    it('should render', () => expect(component.find('button').prop('title')).toBe(description))
  })

  describe('Link', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} link />)
    })

    it('should render', () => {
      const button = component.find('button')
      expect(button.hasClass('my-button-link')).toBeTruthy()
      expect(button.hasClass('my-clickable')).toBeTruthy()
      expect(button.hasClass('my-light-fg')).toBeTruthy()
      expect(button.prop('onClick')).toBe(action)
      expect(button.prop('title')).toBeUndefined()
    })
  })

  describe('Color', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} color="primary" />)
    })

    it('should render', () => expect(component.find('button').hasClass('my-primary-bg')).toBeTruthy())
  })

  describe('Icon', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} icon={icon} />)
    })

    it('should render icon', () => {
      const icon = component.find('Icon')
      expect(icon.prop('icon')).toBeDefined()
      expect(icon.prop('label')).toBe(label)
    })
  })

  describe('Disabled', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} disabled />)
    })

    it('should render', () => expect(component.find('button').hasClass('my-disabled-fg')).toBeTruthy())
  })

  describe('Active', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} active />)
    })

    it('should render', () => expect(component.find('button').hasClass('my-active')).toBeTruthy())
  })

  describe('Class', () => {
    beforeAll(() => {
      component = shallow(<Button action={action} label={label} className={className} />)
    })

    it('should render', () => expect(component.find('button').hasClass(className)).toBeTruthy())
  })

  describe('On click', () => {
    beforeAll(() => {
      clearMocks()
      component = shallow(<Button action={action} label={label} />)

      act(() => {
        component.find('button').simulate('click')
        component.update()
      })
    })

    it('should call action', () => expect(action).toBeCalledTimes(1))
  })
})
