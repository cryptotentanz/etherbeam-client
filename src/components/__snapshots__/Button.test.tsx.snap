// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Default should match snapshot 1`] = `
<button
  class="my-button my-light-bg my-clickable"
>
  <span
    class=""
  >
    Label
  </span>
</button>
`;
